File: vol2/ch18/LeetCode802-2.py
class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        n=len(graph)
        revadj=[[] for _ in range(n)]   #反图邻接表
        degree=[0]*n                    #存放反图的入度
        ans=set()
        qu=deque()                      #对反图进行拓扑排序(队列型)
        for i in range(n):
            degree[i]=len(graph[i])
            if degree[i]==0:qu.append(i)
            for j in graph[i]:
                revadj[j].append(i)
        while qu:
            cnt=len(qu)
            for i in range(0,cnt):
                u=qu.popleft()
                ans.add(u)
                for v in revadj[u]:
                    degree[v]-=1
                    if degree[v]==0:qu.append(v)  
        return sorted(list(ans))

File: vol2/ch18/LeetCode802-2.cpp
class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int n=graph.size();
        vector<vector<int>> revadj(n);//反图邻接表
        vector<int> degree(n,0);               //存放反图的入度
        vector<int> ans;
        queue<int> qu;                      //对反图进行拓扑排序
        for(int i=0;i<n;i++) {
            degree[i]=graph[i].size();
            if(degree[i]==0) qu.push(i);      //入度为0的顶点进队
            for(int j:graph[i])
                revadj[j].push_back(i);
        }
        while(!qu.empty()) {                //分层次的层次遍历
            int cnt=qu.size();
            for(int i=0;i<cnt;i++) {
                int u=qu.front();qu.pop();
                ans.push_back(u);
                for(int v:revadj[u]) {
                    degree[v]-=1;
                    if(degree[v]==0) qu.push(v);
                }
            }
        }
        sort(ans.begin(),ans.end());
        return ans;
    }
};

